File: lib/integrations/credentials.ts
type EncryptedEnvelope = {
  encryptedKey: string;
  iv: string;
  ciphertext: string;
  tag: string;
};

export type CdekCredentials = {
  clientId: string;
  clientSecret: string;
};

export type TelegramCredentials = {
  botToken: string;
};

const textDecoder = new TextDecoder();
const cached = new Map<
  string,
  { credentials: unknown; expiresAt: number }
>();

async function getRuntimeEnv() {
  const { env } = await import("cloudflare:workers");
  return env;
}

function fromBase64(value: string) {
  const binary = atob(value);
  const bytes = new Uint8Array(binary.length);
  for (let index = 0; index < binary.length; index += 1) {
    bytes[index] = binary.charCodeAt(index);
  }
  return bytes;
}

function pemToBytes(pem: string) {
  return fromBase64(
    pem
      .replace("-----BEGIN PRIVATE KEY-----", "")
      .replace("-----END PRIVATE KEY-----", "")
      .replace(/\s/g, ""),
  );
}

async function decryptEnvelope<T>(envelope: EncryptedEnvelope) {
  const env = await getRuntimeEnv();
  const privateKeyPem = env.INTEGRATION_SECRETS_PRIVATE_KEY?.trim();
  if (!privateKeyPem) throw new Error("Ключ защищённых интеграций не настроен");

  const privateKey = await crypto.subtle.importKey(
    "pkcs8",
    pemToBytes(privateKeyPem),
    { name: "RSA-OAEP", hash: "SHA-256" },
    false,
    ["decrypt"],
  );
  const aesKeyBytes = await crypto.subtle.decrypt(
    { name: "RSA-OAEP" },
    privateKey,
    fromBase64(envelope.encryptedKey),
  );
  const aesKey = await crypto.subtle.importKey(
    "raw",
    aesKeyBytes,
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );
  const encrypted = new Uint8Array([
    ...fromBase64(envelope.ciphertext),
    ...fromBase64(envelope.tag),
  ]);
  const clear = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv: fromBase64(envelope.iv), tagLength: 128 },
    aesKey,
    encrypted,
  );
  return JSON.parse(textDecoder.decode(clear)) as T;
}

async function getEncryptedCredentials<T>(provider: "cdek" | "telegram") {
  const current = cached.get(provider);
  if (current && current.expiresAt > Date.now()) {
    return current.credentials as T;
  }
  const env = await getRuntimeEnv();
  const row = await env.DB.prepare(
    "SELECT encrypted_payload FROM integration_credentials WHERE provider = ? LIMIT 1",
  )
    .bind(provider)
    .first<{ encrypted_payload: string }>();
  if (!row?.encrypted_payload) {
    throw new Error(`Доступ к интеграции ${provider} ещё не передан сайту`);
  }

  const credentials = await decryptEnvelope<T>(
    JSON.parse(row.encrypted_payload) as EncryptedEnvelope,
  );
  cached.set(provider, {
    credentials,
    expiresAt: Date.now() + 10 * 60 * 1000,
  });
  return credentials;
}

export async function getCdekCredentials() {
  const credentials = await getEncryptedCredentials<CdekCredentials>("cdek");
  if (!credentials.clientId?.trim() || !credentials.clientSecret?.trim()) {
    throw new Error("Данные доступа к СДЭК неполные");
  }
  return credentials;
}

export async function getTelegramCredentials() {
  const credentials =
    await getEncryptedCredentials<TelegramCredentials>("telegram");
  if (!credentials.botToken?.trim()) {
    throw new Error("Токен Telegram-бота не настроен");
  }
  return credentials;
}

export async function storeEncryptedCredentials(
  provider: "cdek" | "telegram",
  envelope: EncryptedEnvelope,
) {
  const env = await getRuntimeEnv();
  const fields = ["encryptedKey", "iv", "ciphertext", "tag"] as const;
  if (fields.some((field) => !envelope[field] || envelope[field].length > 4096)) {
    throw new Error("Некорректный защищённый пакет");
  }
  await env.DB.prepare(
    `INSERT INTO integration_credentials (provider, encrypted_payload, updated_at)
     VALUES (?, ?, CURRENT_TIMESTAMP)
     ON CONFLICT(provider) DO UPDATE SET
       encrypted_payload = excluded.encrypted_payload,
       updated_at = CURRENT_TIMESTAMP`,
  )
    .bind(provider, JSON.stringify(envelope))
    .run();
  cached.delete(provider);
}

File: backend/internal/integration/credentials.go
package integration

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rsa"
	"crypto/sha256"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"errors"
	"fmt"
	"strings"

	"github.com/jackc/pgx/v5/pgxpool"
)

type CredentialStore struct {
	pool       *pgxpool.Pool
	privateKey *rsa.PrivateKey
}

type encryptedEnvelope struct {
	EncryptedKey string `json:"encryptedKey"`
	IV           string `json:"iv"`
	Ciphertext   string `json:"ciphertext"`
	Tag          string `json:"tag"`
}

func NewCredentialStore(pool *pgxpool.Pool, privateKeyPEM string) (*CredentialStore, error) {
	privateKeyPEM = strings.ReplaceAll(strings.TrimSpace(privateKeyPEM), `\n`, "\n")
	if privateKeyPEM == "" {
		return &CredentialStore{pool: pool}, nil
	}
	block, _ := pem.Decode([]byte(privateKeyPEM))
	if block == nil {
		return nil, errors.New("integration private key is not PEM")
	}
	key, err := x509.ParsePKCS8PrivateKey(block.Bytes)
	if err != nil {
		return nil, fmt.Errorf("parse integration private key: %w", err)
	}
	rsaKey, ok := key.(*rsa.PrivateKey)
	if !ok {
		return nil, errors.New("integration private key is not RSA")
	}
	return &CredentialStore{pool: pool, privateKey: rsaKey}, nil
}

func GetCredentials[T any](
	ctx context.Context,
	store *CredentialStore,
	provider string,
) (T, error) {
	var empty T
	if store.privateKey == nil {
		return empty, errors.New("ключ защищённых интеграций не настроен")
	}

	var payload string
	if err := store.pool.QueryRow(ctx, `
		SELECT encrypted_payload
		FROM integration_credentials
		WHERE provider = $1
		LIMIT 1
	`, provider).Scan(&payload); err != nil {
		return empty, fmt.Errorf("load %s credentials: %w", provider, err)
	}

	var envelope encryptedEnvelope
	if err := json.Unmarshal([]byte(payload), &envelope); err != nil {
		return empty, fmt.Errorf("decode %s envelope: %w", provider, err)
	}
	decode := func(value string) ([]byte, error) {
		return base64.StdEncoding.DecodeString(value)
	}
	encryptedKey, err := decode(envelope.EncryptedKey)
	if err != nil {
		return empty, fmt.Errorf("decode encrypted key: %w", err)
	}
	aesKey, err := rsa.DecryptOAEP(sha256.New(), nil, store.privateKey, encryptedKey, nil)
	if err != nil {
		return empty, fmt.Errorf("decrypt integration key: %w", err)
	}
	iv, err := decode(envelope.IV)
	if err != nil {
		return empty, fmt.Errorf("decode integration iv: %w", err)
	}
	encrypted, err := decode(envelope.Ciphertext)
	if err != nil {
		return empty, fmt.Errorf("decode integration ciphertext: %w", err)
	}
	tag, err := decode(envelope.Tag)
	if err != nil {
		return empty, fmt.Errorf("decode integration tag: %w", err)
	}

	blockCipher, err := aes.NewCipher(aesKey)
	if err != nil {
		return empty, fmt.Errorf("create integration cipher: %w", err)
	}
	gcm, err := cipher.NewGCM(blockCipher)
	if err != nil {
		return empty, fmt.Errorf("create integration gcm: %w", err)
	}
	clear, err := gcm.Open(nil, iv, append(encrypted, tag...), nil)
	if err != nil {
		return empty, fmt.Errorf("decrypt %s credentials: %w", provider, err)
	}

	var credentials T
	if err := json.Unmarshal(clear, &credentials); err != nil {
		return empty, fmt.Errorf("decode %s credentials: %w", provider, err)
	}
	return credentials, nil
}
